### PR TITLE
Ajout des docstrings de build_reports.py

### PR DESCRIPTION
**Pull request by QMaire**

### CODE REVIEW

**Comment by QMaire** [comment]
Seul le fichier build_report.py a été traité.
Les autres arriveront progressivement

File: build_reports.py [hunk 1]
```diff
@@ -201,19 +211,14 @@ def add_weighted_value(pp_dep, pp_reg, pp_nat):
...
-def import_json_to_dict(url) :
-    response = urllib.request.urlopen(url)
-    my_dict = json.loads(response.read())
-    return my_dict
-
-
-def mkdir_ifnotexist(path) :
-    if not os.path.isdir(path) :
+def mkdir_ifnotexist(path: str):
+    """ Create a folder if is not exist"""
```
Comment (desrousseaux) [inline build_reports.py:215]: if it doesn't exist

File: build_reports.py [hunk 2]
```diff
@@ -222,7 +227,8 @@ def format_thousands(s):
     return new_str.strip()
 
 
-def format_amount(indic, valeur):
+def format_amount(indic: str, valeur: str) -> str:
+    """ Add unite after amount """
```
Comment (desrousseaux) [inline build_reports.py:231]: unit

File: build_reports.py [hunk 3]
```diff
@@ -238,19 +244,22 @@ def format_amount(indic, valeur):
             print(f"L'indicateur {indic} possède des valeurs invalides : {err}")
 
 
-def check_pp_reg(pp_reg, taxo_reg_df, pp_dep):
+def check_pp_reg(pp_reg: pd.DataFrame, taxo_reg_df: pd.DataFrame, pp_dep: pd.DataFrame):
+    """ Check if in pp_reg we have all regions and measures """
```
Comment (desrousseaux) [inline build_reports.py:248]: mettre in pp_reg à la fin de la phrase

File: build_reports.py [hunk 4]
```diff
@@ -238,19 +244,22 @@ def format_amount(indic, valeur):
...
     assert sorted(pp_reg['reg'].unique()) == sorted(taxo_reg_df['reg'])
     assert sorted(pp_reg['region'].unique()) == sorted(taxo_reg_df['libelle'])
     assert sorted(pp_reg['mesure'].unique()) == sorted(pp_dep['mesure'].unique())
     assert sorted(pp_reg['short_mesure'].unique()) == sorted(pp_dep['short_mesure'].unique())
     
 
-def check_pp_nat(pp_nat):
+def check_pp_nat(pp_nat: pd.DataFrame):
+    """ Check if in pp_nat we have all measures"""
```
Comment (desrousseaux) [inline build_reports.py:256]: mettre in pp_nat à la fin de la phrase

File: build_reports.py [hunk 5]
```diff
@@ -238,19 +244,22 @@ def format_amount(indic, valeur):
...
     assert sorted(pp_nat['mesure'].unique()) == sorted(pp_nat['mesure'].unique())
     assert sorted(pp_nat['short_mesure'].unique()) == sorted(pp_nat['short_mesure'].unique())
 
 
-def complete_values_for_missing_dates(df_plot, na_replacement, last_dates_to_keep, months):
+def complete_values_for_missing_dates(df_plot: pd.DataFrame, na_replacement: str, last_dates_to_keep: list, months: tuple) ->pd.DataFrame:
+    """ For each date in last_dates_to_keep, if the date is missing in df_plot then we complete with na_replacement """
```
Comment (desrousseaux) [inline build_reports.py:262]: s/complete/fill

File: build_reports.py [hunk 6]
```diff
@@ -272,7 +281,25 @@ def complete_values_for_missing_dates(df_plot, na_replacement, last_dates_to_kee
     return pd.concat([df_plot, df_complement]).reset_index(drop=True)
 
 
-def make_pp_chart(maille, mesure, short_indics, pp_dep, pp_reg, pp_nat, taxo_dep_df, last_dates_to_keep, months, all_charts_as_df):
+def make_pp_chart(maille: str, mesure: str, short_indics: list, pp_dep: pd.DataFrame, pp_reg: pd.DataFrame,
+                  pp_nat: pd.DataFrame, taxo_dep_df: pd.DataFrame, last_dates_to_keep: list, months: tuple, all_charts_as_df: dict):
+    """ For one mesure and one maille, create every tables we need to generate parlementary file, and stock them in all_charts_as_df.
```
Comment (desrousseaux) [inline build_reports.py:286]: every table

File: build_reports.py [hunk 7]
```diff
@@ -272,7 +281,25 @@ def complete_values_for_missing_dates(df_plot, na_replacement, last_dates_to_kee
...
 
-def make_pp_chart(maille, mesure, short_indics, pp_dep, pp_reg, pp_nat, taxo_dep_df, last_dates_to_keep, months, all_charts_as_df):
+def make_pp_chart(maille: str, mesure: str, short_indics: list, pp_dep: pd.DataFrame, pp_reg: pd.DataFrame,
+                  pp_nat: pd.DataFrame, taxo_dep_df: pd.DataFrame, last_dates_to_keep: list, months: tuple, all_charts_as_df: dict):
+    """ For one mesure and one maille, create every tables we need to generate parlementary file, and stock them in all_charts_as_df.  
+    Depending on the parameter maille, tables will be create:
+        - with pp_dep if maille = départemental, for all department
+        - with pp_reg if maille = régional, for all region
+        - pp_nat otherwise
+    Behind, layout of one created table with pp_dep:
```
Comment (desrousseaux) [inline build_reports.py:291]: Below

File: build_reports.py [hunk 8]
```diff
@@ -363,14 +390,19 @@ def make_pp_chart(maille, mesure, short_indics, pp_dep, pp_reg, pp_nat, taxo_dep
...
         short_indics = dict_mesure_indic[mesure]
         for maille in mailles :
             make_pp_chart(maille, mesure, short_indics, pp_dep, pp_reg, pp_nat, taxo_dep_df, last_dates_to_keep, months, all_charts_as_df)
 
 
-def check_charts_exhaustivity(all_charts_as_df, taxo_dep_df, taxo_reg_df, dict_mesure_indic):
+def check_charts_exhaustivity(all_charts_as_df: dict, taxo_dep_df: pd.DataFrame, taxo_reg_df: pd.DataFrame, dict_mesure_indic: dict):
+    """ Multiple checks
+            - Check if in all_charts_as_df, there are all department, regions and the key France
+            - Check if in all_charts_as_df, there are all tables for each department, regions and for the key France"""
```
Comment (desrousseaux) [inline build_reports.py:405]: Check if all charts_as_df contains ...

File: build_reports.py [hunk 9]
```diff
@@ -383,7 +415,14 @@ def check_charts_exhaustivity(all_charts_as_df, taxo_dep_df, taxo_reg_df, dict_m
     
     assert sorted(all_charts_as_df['national']['France'].keys()) == sorted(dict_mesure_indic.keys())
 
-def mesure_to_insert(pp_dep):
+def mesure_to_insert(pp_dep: pd.DataFrame) -> list:
+    """ Business fonction. Only some measures will be present in parlementary file and for the latter
```
Comment (desrousseaux) [inline build_reports.py:419]: On peut enlever le Business function je pense

File: build_reports.py [hunk 10]
```diff
@@ -444,7 +483,8 @@ def mesure_to_insert(pp_dep):
     return short_mesure2url, short_mesure2to_comment
 
 
-def create_front_page(nom_departement):
+def create_front_page(nom_departement: str) -> str:
+    """ Create for one departement, the front page of the parlementary file """
```
Comment (desrousseaux) [inline build_reports.py:487]: Given a departement, creates the front ...

File: build_reports.py [hunk 11]
```diff
@@ -467,25 +508,28 @@ def encode_name(name):
     return name
 
 
-def create_volet_page(nom_volet, num_volet):
+def create_volet_page(nom_volet: str, num_volet: int) -> str:
+    """ Create for one departement, the "volet" page of the parlementary file """
```
Comment (desrousseaux) [inline build_reports.py:512]: Given a departement, creates...

File: build_reports.py [hunk 12]
```diff
@@ -496,7 +540,9 @@ def fusion_word(word1, word2, dep):
     return name_fusion
 
 
-def create_content_page(all_charts_as_df, departement, region, mesure, volet, dep_name, reg_name, num_mesure, short_mesure2to_comment, short_mesure2url, dict_mesure_indic):
+def create_content_page(all_charts_as_df: dict, departement: str, region: str, mesure: str, volet: str, dep_name: str, reg_name: str,
+                        num_mesure: int, short_mesure2to_comment: dict, short_mesure2url: dict, dict_mesure_indic: dict) -> str:
+    """ Create one page for parlementary file. This page is composed of 3 tables (departmental, regional, national) for one measure  """
```
Comment (desrousseaux) [inline build_reports.py:545]: This page is made of...

**Comment by QMaire** [comment]
build_reports.py corrigé après Review
transpose_comments.py ajouté

File: build_reports.py [hunk 13]
```diff
@@ -122,7 +119,10 @@ def main_build_reports():
 
 # Fonction nécessaire
 
-def create_dict_mesure_indic(pp_dep, volet2code_mesures):
+def create_dict_mesure_indic(pp_dep: pd.DataFrame, volet2code_mesures: dict) -> dict:
+    """
+    Create crossing dictionnary between Mesure (key) and Indicateur (value)
```
Comment (vesran) [inline build_reports.py:124]: " CreateS crossing ..."
The function creates a crossing dictionary.

File: build_reports.py [hunk 14]
```diff
@@ -165,7 +175,10 @@ def add_cumulated_value(pp_dep, taxo_reg_df):
     return pp_reg, pp_nat
 
 
-def months_to_insert(modulo = 0, months=months, nb_mois=3):
+def months_to_insert(modulo: int = 0, months: tuple = months, nb_mois: int = 3) -> list:
```
Comment (vesran) [inline build_reports.py:178]: WOuld it be better to say 
```
def insert_months_to(...)
```
to keep the structure verb-complement ?

File: build_reports.py [hunk 15]
```diff
@@ -272,7 +303,26 @@ def complete_values_for_missing_dates(df_plot, na_replacement, last_dates_to_kee
...
+    For one mesure and one maille, create every table we need to generate parlementary file, and stock them in all_charts_as_df.  
+    Depending on the parameter maille, tables will be create:
+        - with pp_dep if maille = départemental, for all department
+        - with pp_reg if maille = régional, for all region
+        - pp_nat otherwise
+    Below, layout of one created table with pp_dep: 
+     
+            -----------------------
+            |        mesure        |
+            ------------------------
```
Comment (vesran) [inline build_reports.py:318]: Not sure whether mesure is a dataframe header or not.

File: docx2pdf.py [hunk 16]
```diff
@@ -38,20 +38,26 @@ def main_docx2pdf_apres_osmose():
...
+    Create a folder if it's doesn't exist
+    """
+    if not os.path.isdir(path):
         os.mkdir(path)
         
 
-def create_dico_dep2name(taxo_dep_df):
+def create_dico_dep2num(taxo_dep_df: pd.DataFrame) -> dict:
+    """
+    Create crossing dictionnary between department and department's number
```
Comment (vesran) [inline docx2pdf.py:59]: Creates crossing dictionnary between department's **name** and department's number

File: main_convert_parlementary_file_with_new_comment.py [hunk 17]
```diff
@@ -24,12 +24,17 @@ def main():
...
+    Create a folder if it's doesn't exist
+    """
+    if not os.path.isdir(path):
         os.mkdir(path)
 
-def create_zip_for_archive(name_zip, folder_pdf, folder_docx):
-    #Code permettant de créer un zip avec l'arborescence du fichier source
+def create_zip_for_archive(name_zip: str, folder_pdf: str, folder_docx: str):
+    """
+    Create a zip in archive/Month_Year with 2 folders: folder_pdf and forlder_docx
```
Comment (vesran) [inline main_convert_parlementary_file_with_new_comment.py:36]: Creates a zip in archive/Month_Year with 2 folders: folder_pdf and folder_docx

File: main_create_parlementary_file.py [hunk 18]
```diff
@@ -40,18 +41,27 @@ def main():
...
+    Create a folder if it's doesn't exist
+    """
+    if not os.path.isdir(path):
         os.mkdir(path)
 
-def create_zip_for_archive(name_zip, folder_pdf, folder_docx):
-    #Code permettant de créer un zip avec l'arborescence du fichier source
+def create_zip_for_archive(name_zip: str, folder_pdf: str, folder_docx: str):
+    """
+    Create a zip in archive/Month_Year with 2 folders: folder_pdf and forlder_docx
```
Comment (vesran) [inline main_create_parlementary_file.py:63]: Creates a zip in archive/Month_Year with 2 folders: folder_pdf and folder_docx

File: transpose_comments.py [hunk 19]
```diff
@@ -86,63 +88,98 @@ def main_transpose_comments():
...
             yield from flatten(item)
     else:
         yield L
         
 
-def gen_unit_list(L):
-    # Générer 
+def gen_unit_list(L: list):
+    """
+    ???
```
Comment (vesran) [inline transpose_comments.py:129]: Creates a generator that generates the smallest list for nested lists : [ [[.], [., .], [.]], [ [a], [.] ] ] --> [.], [., .], [.], [a], [.]

File: transpose_comments.py [hunk 20]
```diff
@@ -184,7 +221,13 @@ def extract_comment(textbox_content):
     return textbox_content
     
 
-def alternate_texts_and_images(doc, textbox_content):
+def alternate_texts_and_images(doc: DocxTemplate, textbox_content: str) -> list:
+    """
+    Comments may contain pictures. To keep them, we create a liste to store text and image path
+
```
Comment (vesran) [inline transpose_comments.py:227]: ... The image size is fixed since it cannot be extracted from the initial document.

**Review by vesran — CHANGES_REQUESTED** [review]

**Review by vesran — APPROVED** [review]

**Pull request merged**